Add support for koajs.

gtarcea opened the pull request:
The following request adds support for koajs and I suspect other app servers though I have not tested it. The changes are minor. In the case of koajs, we pass in the http server, there is no 'app.on' function hanging off of app._events.request. In this case you want to use the original server passed in.

This change explicitly checks if 'on' is hanging off of app, or if we should use the server passed in. I've tested this in my project using koa. Please let me know if there are any changes you would like to see.

alaa-eddine commented:
I'll check if there is no conflict with expressjs and merge it if everything is OK :)

this part of the code need some cleaner way to detect frameworks/middlewares, I don't like the way I've did it :D but it work.

maybe in a next version I'll use a wrapper class to detect frameworks like I do with transport layers.

thank you for the contribution

alaa-eddine commented:
btw, I never used koajs, can you please post a simple code to demonstrate the usage ?
cause I'll need to add an exemple code for this

gtarcea commented:
I'll put together an example that follows your simple server example that used express (with koajs instead). The differences are really small. If I have time I'll try this out with restify as well. I'm hoping my change means it will support additional platforms.

Would you accept a patch that added promise to support? Then instead of xx.hello().onReady(function()...), you could use xx.hello().then(function()...). The big benefit would be with es6. I could do:
var results = yield xx.hello();

This is a really great project!

alaa-eddine commented:
thank you 
you can add the example directly to the pull request so I'll merge it with the examples.

put it under folder examples/9-koajs/

for the promises I already started working on them, it's still not available on npm but it's already present in the git repo here https://github.com/Ezelia/eureca.io/blob/master/src/Stub.ts.

I didn't used a full promises library and the actual implementation will not return all failing messages but it's working.

obviously if you have suggestions to enhance it you're  welcome :)

alaa-eddine commented:
@gtarcea I'm trying to understand the modification you made :)

```
(1)            var requestHandler = server; 
(2)            if (server._events && server._events.request !== undefined && server.routes === undefined) app = server._events.request;
(3) if (app.on) requestHandler = app;
```

what is the purpose of line (3) above ? 
because line (2) is already checking if the server is bound to an application ? if so app.on is used, otherwise original server is used.

gtarcea commented:
I'll add the examples and send an update to the pull request.

The code is a little odd. I probably should have added a comment. I'll put one in with the update. This is probably best explained by showing some example code using koajs.

I've left out some details, but here are the basics of setting up koajs to use eureca.
```javascript
var app = require('koa')();
var eurecaServer = new Eureca.Server();
var server = require('http').createServer(app.callback());
eurecaServer.attach(server);
```
Ok, so at this point the following check is true
```javascript
if (server._events && server._events.request !== undefined && server.routes === undefined)
    app = server._events.request;
```
Here is the problem. There is no `on` method exposed on `server._events.request`. That means that later on in the code this call will get an undefined:
```javascript
onProvider.on('request', function (request, response) {
```
Since koajs isn't exposing `on` on the request, we really need to get it off of the base server that was passed in. However, express does expose the `on` method on request, so I needed to handle that case. What I do is explicitly check if `on` is supported on the app, and if not use the server instead. So this check:
```javascript
var requestHandler = server;
....
if (app.on) requestHandler = app;
```
Explicitly checks if there is an `on` method on the app. If there isn't then we need to use the server, which I assigned to the requestHandler.

Let me know if that helps.

alaa-eddine commented:
I understand the check of app.on but isn't it simpler to just add the check to the above line so the original code becomes : 

```javascript
            var app = server;
            if (server._events.request !== undefined && server.routes === undefined && server._events.request.on) app = server._events.request;
```

I'm maybe missing something, but if the need here is only to check the existance of on, it can be done that way :)

this is maybe not the best method to detect express/koajs but I don't think that those frameworks  provide a clean way to detect their presence .

gtarcea commented:
Hi @alaa-eddine ,
I thought of that as well. The reason I didn't take that approach is I don't know how to test this section of the code in attach (that is I don't know how to cause that part of the code to activate):
```javascript
//install on express
//sockjs_server.installHandlers(server, {prefix:_prefix});
            if (app.get && app.post)  //TODO : better way to detect express
            {
                app.get(_clientUrl, function (request, response) {
                    _this.sendScript(request, response, _prefix);
                });
            }
```
So, I hedged my bet in case this was important, and not something that the http server would do. 

If the `if` statement is updated it'll need to check `server._events` before executing the rest of the `if` tests, otherwise it will fail for the koajs case. So the if becomes:
```javascript
var app = server;
if (server._events && server._events.request !== undefined && server.routes === undefined && server._events.request.on) app = server._events.request;
```
I'll try and send you an update to the pull request with the examples tonight, but more likely tomorrow (Thursday). The flu has decided to pay my family a visit :-|

alaa-eddine commented:
haha, no problem take your time  
I wish you a quick recovery :)

gtarcea commented:
I've updated the pull request with an example for koajs. I also simplified the checking for assigning app using your suggestions above. Let me know if there are any changes you would like made to the example. I tried to make it comprehensive by showing the client calling the server, and the server calling the client.

Also I tried testing with restify. I wasn't able to get it working. It looks like Restify is intercepting the calls in unexpected spots. After eureca writes header/status Restify tries to resulting in an exception. I'll probably try HAPI just to see if that works. From my perspective koa was the important server to support.

Thanks for the well wishes for the flu. Everyone seems to be recovering :-)

alaa-eddine commented:
Merged :)

I need to enhance this part of code to make it compatible with all such frameworks.
it's only needed to send eureca.js and it's a bad thing that it make the whole eureca.io incompatible with many frameworks just for this.

alaa-eddine commented:
just published the npm package including your changes and an implementation of promises.

feel free to make new PR if you think the promises implementation can be enhanced.

gtarcea commented:
Thank you so much. This is wonderful. I'll play around with promises.

I intend on using this framework on a couple of projects. If there is 
anything I can do to help this work across other frameworks let me know.

> Alaa-eddine K. <mailto:notifications@github.com>
> April 26, 2015 at 8:26 AM
>
> just published the npm package including your changes and an 
> implementation of promises.
>
> feel free to make new PR if you think the promises implementation can 
> be enhanced.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/Ezelia/eureca.io/pull/14#issuecomment-96372463>.
>
